core/include/gnuradio-4.0/Scheduler.hpp

```diff
@@ -896,20 +928,108 @@ detecting cycles and blocks which can be reached from several source blocks.)"">
-            auto& job = this->_jobLists->emplace_back(std::vector<BlockModel*>());
-            job.reserve(_blocklist.size() / n_batches + 1);
-            for (std::size_t j = i; j < _blocklist.size(); j += n_batches) {
-                job.push_back(_blocklist[j]);
+            auto& job = this->_executionOrder->emplace_back(std::vector<BlockModel*>());
+            job.reserve(blockList.size() / n_batches + 1);
+            for (std::size_t j = i; j < blockList.size(); j += n_batches) {
+                job.push_back(blockList[j]);
+            }
+        }
+    }
+
+    void reset() {
+        base_t::reset();
+        init();
+    }
+};
+
+template<ExecutionPolicy execution = ExecutionPolicy::singleThreaded, profiling::ProfilerLike TProfiler = profiling::null::Profiler>
+class DepthFirst : public SchedulerBase<DepthFirst<execution, TProfiler>, execution, TProfiler> {
+    using Description = Doc<R""(Depth First Scheduler which traverses the graph starting from the source blocks in a depth-first manner.)"">;
+
+    friend class lifecycle::StateMachine<DepthFirst<execution, TProfiler>>;
+    friend class SchedulerBase<DepthFirst<execution, TProfiler>, execution, TProfiler>;
+    static_assert(execution == ExecutionPolicy::singleThreaded || execution == ExecutionPolicy::multiThreaded, "Unsupported execution policy");
+
+public:
+    using base_t = SchedulerBase<DepthFirst<execution, TProfiler>, execution, TProfiler>;
+
+    explicit DepthFirst(gr::Graph&& graph, std::string_view defaultThreadPool = gr::thread_pool::kDefaultCpuPoolId, const profiling::Options& profiling_options = {}) : base_t(std::move(graph), defaultThreadPool, profiling_options) {}
+    explicit DepthFirst(gr::Graph&& graph, property_map initialSettings, std::string_view defaultThreadPool = gr::thread_pool::kDefaultCpuPoolId) : base_t(std::move(graph), std::move(initialSettings), defaultThreadPool) {}
+
+private:
+    void init() {
+        std::vector<BlockModel*>    blocklist;
+        [[maybe_unused]] const auto pe = this->_profilerHandler.startCompleteEvent("depth_first.init");
+        using block_t                  = BlockModel*;
+        base_t::init();
+
+        // adjacency list
+        std::map<block_t, std::vector<block_t>> _adjacency_list{};
+        std::vector<block_t>                    _source_blocks{};
+        std::set<block_t>                       block_reached;
+
+        for (auto& e : this->_graph.edges()) {
+            _adjacency_list[e._sourceBlock].push_back(e._destinationBlock);
+            _source_blocks.push_back(e._sourceBlock);
+            block_reached.insert(e._destinationBlock);
+        }
+
+        _source_blocks.erase(std::remove_if(_source_blocks.begin(), _source_blocks.end(), [&block_reached](auto currentBlock) { return block_reached.contains(currentBlock); }), _source_blocks.end());
+
+        // depth first traversal — use stack
+        std::set<block_t> visited;
+
+        std::function<void(block_t)> dfs = [&](block_t node) {
```

Comment:
We can now use `this auto self` for recursive lambdas for better speed than std::function. (or, this can be a Y combinator-inspired function and take itself as the first argument for older C++ compilers).